builder/dockerfile/internals.go

```diff
@@ -371,38 +359,56 @@ func (b *Builder) calcCopyInfo(cmdName, origPath string, allowLocalDecompression
 	}
 
 	// Must be a dir or a file
-	statPath, fi, err := context.Stat(origPath)
+	hash, err := source.Hash(origPath)
 	if err != nil {
 		return nil, err
 	}
 
-	copyInfos := []copyInfo{{FileInfo: fi, decompress: allowLocalDecompression}}
-
-	hfi, handleHash := fi.(builder.Hashed)
-	if !handleHash {
-		return copyInfos, nil
+	fi, err := remotecontext.StatAt(source, origPath)
+	if err != nil {
+		return nil, err
 	}
+
+	// TODO: remove, handle dirs in Hash()
```

Comment:
Yes, currently `Hash()` only returns hash for a single file so to get the correct hash for a dir we still need to recursively walk it. The todo is about moving the walker code inside `Hash` so the values from that func don't need post-processing.